plugin/src/main/java/org/openrewrite/gradle/RewritePlugin.java

```diff
@@ -99,9 +99,8 @@ private static void configureProject(Project project, RewriteExtension extension
 
             //Collect Java metadata for each project (used for Java Provenance)
             //Using the older javaConvention because we need to support older versions of gradle.
-            @SuppressWarnings("deprecation")
-            JavaPluginConvention javaConvention = project.getConvention().getPlugin(JavaPluginConvention.class);
-            javaConvention.getSourceSets().all(sourceSet -> {
+            JavaPluginExtension javaPluginExtension = project.getExtensions().getByType(JavaPluginExtension.class);
+            javaPluginExtension.getSourceSets().all(sourceSet -> {
```

Comment:
We tried this in the past, but it looks like `getSourceSets()` was not added until 7.1 so we reverted the change again: https://github.com/openrewrite/rewrite-gradle-plugin/pull/281#issuecomment-1970694368

However, we will also want to be Gradle 9 compatible, so we might want to use both APIs here, depending on the Gradle version.

What do you think @shanman190 ?

Reply:
Yeah, that'll have to be what we do. What I would probably do is hide the complexity in a `SourceSetContainer getSourceSets()` method. For pre-7.1, we'd use reflection to load the convention class, get the instance, then get the `SourceSetContainer` off of it. Then for 7.1 and later, we'd just code directly to the `JavaPluginExtension` directly. This would allow us to remain backwards compatible as well as being forward compatible to whenever Gradle finally decides to remove the `JavaPluginConvention` class (it's been deprecated since the 7.x line where it was originally scheduled to be removed at 8.0, but was delayed).

Reply:
We can mitigate it by checking which version is running to decide execution of code path, but there is another problem, it breaks test `rewriteRun applies recipe to a multi-project build(File)`:
```
org.opentest4j.AssertionFailedError: 
expected: 
  "package com.foo;
  
  import org.junit.Test;
  
  public class ATestClass {
  
      @Test
      public void passes() {
      }
  }"
 but was: 
  "package com.foo;
  
  import org.junit.Test;
  
  public class ATestClass {
  
      @Test
      public void passes() { }
  }"
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at org.openrewrite.gradle.RewriteRunTest.rewriteRun applies recipe to a multi-project build(RewriteRunTest.kt:255)
```

Reply:
Not entirely sure what the issue is here, but I am sure that could be fixed.

Reply:
@quaff If you would like to contribute a PR which allows the build plugin to use `JavaPluginExtension` rather than `JavaPluginConvention` in Gradle 7+ environments, that would definitely be most welcome ❤️

Reply:
I will give a shot if I'm spare, but I think this PR should focus on `Project.getBuildDir()`.